activesupport/lib/active_support/cache/serializer_with_fallback.rb

```diff
@@ -40,6 +46,40 @@ def load(dumped)
       end
 
       private
+        BARE_STRING_SIGNATURES = {
+          255 => Encoding::UTF_8,
+          254 => Encoding::BINARY,
+          253 => Encoding::US_ASCII,
+        }
+        BARE_STRING_TEMPLATE = "CEl<"
+        BARE_STRING_VERSION_INDEX = [0, 0.0, 0].pack(BARE_STRING_TEMPLATE).bytesize
+
+        def try_dump_bare_string(entry)
+          if entry.value.instance_of?(String) && signature = BARE_STRING_SIGNATURES.key(entry.value.encoding)
+            packed = [
+              signature,
+              entry.expires_at || -1.0,
+              entry.version&.bytesize || -1,
+            ].pack(BARE_STRING_TEMPLATE)
+
+            packed << entry.version if entry.version
+            packed << entry.value
+          end
+        end
+
+        def try_load_bare_string(dumped)
+          if encoding = BARE_STRING_SIGNATURES[dumped.getbyte(0)]
+            _, expires_at, version_length = dumped.unpack(BARE_STRING_TEMPLATE)
```

Comment:
Not sure if worth it, but you could save that array allocation by using `unpack1` with `@` directives, e.g. 

```ruby
>> [255, 23.45, 10].pack("CEl<")
=> "\xFF33333s7@\n\x00\x00\x00"
>> [255, 23.45, 10].pack("CEl<").unpack1("@1E")
=> 23.45
>> [255, 23.45, 10].pack("CEl<").unpack1("@9l<")
=> 10
```

Reply:
`unpack1` gives ~5% improvement:

<details>
  <summary>Benchmark Script</summary>

  ```ruby
  # frozen_string_literal: true

  require "benchmark/ips"

  module ActiveSupport::Cache::SerializerWithFallback::Marshal71WithFallback
    def load_with_unpack1(dumped)
      if dumped.is_a?(String)
        dumped = decompress(dumped) if compressed?(dumped)

        case
        when loaded = try_load_bare_string_with_unpack1(dumped)
          loaded
        else
          raise
        end
      end
    end

    private
      BARE_STRING_UNPACK_EXPIRES_AT = "@1E"
      BARE_STRING_UNPACK_VERSION_LENGTH = "@9l<"

      def try_load_bare_string_with_unpack1(dumped)
        if encoding = BARE_STRING_SIGNATURES[dumped.getbyte(0)]
          expires_at = dumped.unpack1(BARE_STRING_UNPACK_EXPIRES_AT)
          version_length = dumped.unpack1(BARE_STRING_UNPACK_VERSION_LENGTH)
          value_index = BARE_STRING_VERSION_INDEX + [version_length, 0].max

          ActiveSupport::Cache::Entry.new(
            dumped.byteslice(value_index..-1).force_encoding(encoding),
            version: dumped.byteslice(BARE_STRING_VERSION_INDEX, version_length),
            expires_at: (expires_at unless expires_at < 0),
          )
        end
      end
  end


  serializer = ActiveSupport::Cache::SerializerWithFallback[:marshal_7_1]
  entry = ActiveSupport::Cache::Entry.new(Random.bytes(10_000), version: "123")
  dumped = serializer.dump(entry)

  Benchmark.ips do |x|
    x.report("w/ unpack") do
      serializer.load(dumped)
    end

    x.report("w/ unpack1") do
      serializer.load_with_unpack1(dumped)
    end

    x.compare!
  end
  ```
</details>


  ```
  Warming up --------------------------------------
             w/ unpack    27.513k i/100ms
            w/ unpack1    28.810k i/100ms
  Calculating -------------------------------------
             w/ unpack    276.348k (± 1.2%) i/s -      1.403M in   5.078304s
            w/ unpack1    286.904k (± 1.4%) i/s -      1.440M in   5.021905s

  Comparison:
            w/ unpack1:   286904.0 i/s
             w/ unpack:   276348.5 i/s - 1.04x  (± 0.00) slower
  ```

Though I prefer the conciseness of `unpack`, I suppose the reduced memory pressure is worth it, so I've switched to `unpack1`.